c10/core/DispatchKey.h

```diff
@@ -19,48 +27,94 @@ enum class DispatchKey : uint8_t {
   // and Undefined == nullopt.  We didn't actually represent
   // it this way because optional<RealDispatchKey> would take two
   // words, when DispatchKey fits in eight bits.
+
   Undefined = 0,
 
-  // This pool of IDs is not really ordered, but it is merged into
-  // the hierarchy for convenience and performance
-  CPUTensorId, // PyTorch/Caffe2 supported
-  CUDATensorId, // PyTorch/Caffe2 supported
-  MKLDNNTensorId, // Caffe2 only
-  OpenGLTensorId, // Caffe2 only
-  OpenCLTensorId, // Caffe2 only
-  IDEEPTensorId, // Caffe2 only
-  HIPTensorId, // PyTorch/Caffe2 supported
-  SparseHIPTensorId, // PyTorch only
-  MSNPUTensorId, // PyTorch only
-  XLATensorId, // PyTorch only
-  MkldnnCPUTensorId,
-  QuantizedCPUTensorId, // PyTorch only
-  ComplexCPUTensorId, // PyTorch only
-  ComplexCUDATensorId, // PyTorch only
-
-  // See Note [Private use TensorId]
-  PrivateUse1_TensorId,
-  PrivateUse2_TensorId,
-  PrivateUse3_TensorId,
 
-  // Sparse has multi-dispatch with dense; handle it first
-  SparseCPUTensorId, // PyTorch only
-  SparseCUDATensorId, // PyTorch only
 
-  // Custom pseudorandom number generator dispatch key.
-  // To enable PyTorch users to have custom PRNGs, we added support for
-  // dispatching on at::Generator* parameters.
-  // This key must be used in two places:
-  //  1) as a second parameter of at::Generator constructor call in
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~ BACKENDS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
+  // A "backend" is colloquially used to refer to handlers for dispatch
+  // which actually implement the numerics of an operation in question.
+  //
+  // Due to the nature of the enum, these backends are specified in
+  // an ordered way, but for most backends this order is not semantically
+  // meaningful (e.g., it's valid to reorder these backends without changing
+  // semantics).  The only situation when backend ordering is meaningful
+  // is when the backend participates in multiple dispatch with another
+  // backend; e.g., CPUTensorId and SparseCPUTensorId (sparse must have
+  // higher priority).
+
+  // Here are backends which you think of as traditionally specifying
+  // how to implement operations on some device.
+  CPUTensorId,    // registered at build/aten/src/ATen/CPUType.cpp
```

Comment:
not sure I'd mention the `Type.cpp` files. They're supposed to die and then we have this lingering comment here. Also, the ideal final state would be that CPU or CUDA tensors don't have one central file they're defined in anymore.

Reply:
It's mostly to help people trying to navigate the code right now. We can update it when things change.